### ISSUE TITLE
InsufficientSubnetSize error when deploying AKS cluster with /24 subnet

### ISSUE THREAD

**Issue by dnoland1** [rancher/rancher#22616]
**What kind of request is this (question/bug/enhancement/feature request):**
bug

**Steps to reproduce (least amount of steps as possible):**
1. Create a new subnet (used az cli):
```
az network vnet create \
    --resource-group $RS_GROUP \
    --name $VNET_NAME \
    --address-prefixes 10.0.0.0/8 \
    --subnet-name $SUBNET_NAME \
    --subnet-prefix 10.164.15.0/24
```
2. In Rancher UI, create a new AKS cluster. On the second screen, select advanced networking and enter:
Virtual Network - Subnet (Subnet CIDR): 10.164.15.0/24
Kubernetes service address range" 10.13.0.0/16
Kubernetes DNS service IP address: 10.13.0.10
Docker bridge address: 10.11.0.1/16

**Result:**
Cluster fails to provision and get error: `Containerservice.ManagedClustersClient#CreateOrUpdate: Failure sending request: StatusCode=0 -- Original Error: autorest/azure: Service returned an error. Status=400 Code="InsufficientSubnetSize" Message="Pre-allocated IPs 333 exceeds IPs available in Subnet 251" Target="agentPoolProfile.count"`

According to docs at: https://docs.microsoft.com/en-us/azure/aks/configure-kubenet#ip-address-availability-and-exhaustion

> kubenet - a simple /24 IP address range can support up to 251 nodes in the cluster (each Azure virtual network subnet reserves the first three IP addresses for management operations)

**Other details that may be helpful:**
Seems to be same as this issue which was closed by the user: https://github.com/rancher/rancher/issues/20784

Perhaps this is caused by the inability to pass the pod cidr when creating the cluster.

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI):
v2.2.8
- Installation option (single install/HA):
single

**Cluster information**
- Cluster type (Hosted/Infrastructure Provider/Custom/Imported):
AKS
- Machine type (cloud/VM/metal) and specifications (CPU/memory):
VM size `Standard_B2s`
- Kubernetes version (use `kubectl version`):

```
1.14
```

- Docker version (use `docker version`):
na

**Comment by leodotcloud** [comment]
Created a cluster with:
```
export RS_GROUP=leo-resource-group-1
export VNET_NAME=leo-vnet-1
export SUBNET_NAME=leo-subnet-1
```

```
az network vnet create \
    --resource-group $RS_GROUP \
    --name $VNET_NAME \
    --address-prefixes 10.0.0.0/8 \
    --subnet-name $SUBNET_NAME \
    --subnet-prefix 10.164.15.0/24
```

```
"virtualNetwork": "leo-vnet-1",
"virtualNetworkResourceGroup": "leo-resource-group-1"
"serviceCidr": "10.13.0.0/16"
"dnsServiceIp": "10.13.0.10"
"dockerBridgeCidr": "172.17.0.1/16"
"podCidr": "172.244.0.0/16"
```

The cluster was created successfully. The only difference is that I did not specify the docker bridge address in the 10.x.x.x subnet.

**Comment by leodotcloud** [comment]
**Root cause:** 
Even Advanced Networking is selected, the network Plugin was set to `azure` not `kubenet`. 

**Workaround:** Use the API or CLI to create the cluster.

Example:
```
curl -u "${CATTLE_ACCESS_KEY}:${CATTLE_SECRET_KEY}" \
-X POST \
-H 'Accept: application/json' \
-H 'Content-Type: application/json' \
-d '{
    "amazonElasticContainerServiceConfig": null,
    "azureKubernetesServiceConfig": {
        "aadClientAppId": "",
        "aadServerAppId": "",
        "aadServerAppSecret": "",
        "aadTenantId": "",
        "adminUsername": "azureuser",
        "agentDnsPrefix": "",
        "agentOsdiskSize": 100,
        "agentPoolName": "rancher",
        "agentStorageProfile": "ManagedDisks",
        "agentVmSize": "Standard_D2_v2",
        "authBaseUrl": "https://login.microsoftonline.com/",
        "baseUrl": "https://management.azure.com/",
        "clientId": "",
        "clientSecret": "",
        "count": 3,
        "displayName": "",
        "dnsServiceIp": "10.13.0.10",
        "dockerBridgeCidr": "172.17.0.1/16",
        "driverName": "azurekubernetesservice",
        "enableHttpApplicationRouting": false,
        "enableMonitoring": true,
        "kubernetesVersion": "1.14.6",
        "location": "westus",
        "logAnalyticsWorkspace": "",
        "logAnalyticsWorkspaceResourceGroup": "",
        "masterDnsPrefix": "",
        "maxPods": 110,
        "name": "",
        "networkPlugin": "kubenet",
        "networkPolicy": "",
        "podCidr": "172.244.0.0/16",
        "resourceGroup": "leo-resource-group-1",
        "serviceCidr": "10.13.0.0/16",
        "sshPublicKeyContents": "",
        "subnet": "leo-subnet-1",
        "subscriptionId": "",
        "tags": [],
        "tenantId": "",
        "type": "azureKubernetesServiceConfig",
        "virtualNetwork": "leo-vnet-1",
        "virtualNetworkResourceGroup": "leo-resource-group-1"
    },
    "dockerRootDir": "/var/lib/docker",
    "enableClusterAlerting": false,
    "enableClusterMonitoring": false,
    "enableNetworkPolicy": false,
    "googleKubernetesEngineConfig": null,
    "huaweiEngineConfig": null,
    "localClusterAuthEndpoint": null,
    "name": "leo-test-aks-cluster-3",
    "rancherKubernetesEngineConfig": null
}' \
'https://<rancher-server-url>/v3/clusters'
```

**Possible fix:** The UI can provide a toggle button with the network plugin options azure/kubenet.

**Comment by dnoland1** [comment]
That's interesting. When I deploy an AKS cluster using Rancher UI, it seems to be set to kubenet on the Azure side:
```
# az aks list
...
    "networkProfile": {
      "dnsServiceIp": "10.0.0.10",
      "dockerBridgeCidr": "172.17.0.1/16",
      "loadBalancerSku": "basic",
      "networkPlugin": "kubenet",
      "networkPolicy": null,
      "podCidr": "10.244.0.0/16",
      "serviceCidr": "10.0.0.0/16"
    },
...
```

**Comment by deniseschannon** [comment]
Can we add a toggle button with the network plugin options `azure` or `kubenet`?

**Comment by alena1108** [comment]
Along with the toggle button, we have to make sure that we support/pass all the other params depending on the network plugin choice (podCIDR?). @dnoland1 @leodotcloud can you investigate if all of them are available in the UI

**Comment by codyrancher** [comment]
@alena1108 am I right in assuming that the following is a list of all the params we need to be concerned with?
- dnsServiceIp
- dockerBridgeCidr
- loadBalancerSku
- networkPlugin
- networkPolicy
- podCidr
- serviceCidr

If so the UI is missing loadBalancerSku, podCidr and of course networkPlugin. If that's not the case would you mind pointing me in the right direction to find that list?

**Comment by codyrancher** [comment]
In my PR I've added support for:
-dnsServiceIp
-dockerBridgeCidr
-~~loadBalancerSku~~
-networkPlugin
-networkPolicy
-podCidr
-serviceCidr

I removed loadBalancerSku as it doesn't appear to be accepted by the backend. If there are any other fields that need to be added please let me know. I also decided to go with a select instead of a toggle since a networkPlugin doesn't necessarily have to be a binary option.

Here are some screenshots in case there's some feedback.
![Screen Shot 2019-10-15 at 3 28 09 PM](https://user-images.githubusercontent.com/55104481/66874638-a83d4f80-ef60-11e9-8ccd-e03a494e2223.png)
![Screen Shot 2019-10-15 at 3 29 32 PM](https://user-images.githubusercontent.com/55104481/66874637-a83d4f80-ef60-11e9-9746-f2782db3cd3d.png)

**Comment by westlywright** [comment]
available in `latest-2.3` UI tag

**Comment by soumyalj** [comment]
Reproduced the issue with v2.3.0. 
1. Create a subnet using CLI as below:
az network vnet create \
    --resource-group $RS_GROUP \
    --name $VNET_NAME \
    --address-prefixes 10.0.0.0/8 \
    --subnet-name $SUBNET_NAME \
    --subnet-prefix 10.164.15.0/24
2. In Rancher UI, create a new AKS cluster. On the second screen, select advanced networking and enter:
Virtual Network - Subnet (Subnet CIDR): 10.164.15.0/24
Kubernetes service address range" 10.13.0.0/16
Kubernetes DNS service IP address: 10.13.0.10
Docker bridge address: 10.11.0.1/16

Cluster provisioning fails with the below error:
<img width="1278" alt="Screen Shot 2019-10-21 at 12 20 23 PM" src="https://user-images.githubusercontent.com/18536626/67236090-5fbce080-f3fd-11e9-954a-53fda7150790.png">
